src/v/cluster/self_test_frontend.cc

```diff
@@ -0,0 +1,263 @@
+    start_test_request req{.id = id, .dto = dto, .nto = nto};
+    co_await invoke_on_all_nodes(
+      [req](auto handle) { return handle->start_test(req); });
+    /// TODO: Throw if something bad occurred
+    co_return id;
+}
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
+    while (!test_state.finished() && retries-- > 0) {
+        /// invoke_on_all will filter brokers that have shut down cleanly from
+        /// its participants list, retries (if occur) will only occur on
+        /// previously unreachable nodes
+        test_state = co_await invoke_on_all_nodes(
+          [](auto handle) { return handle->stop_test(); });
+        if (!test_state.finished()) {
+            vlog(
+              clusterlog.warn,
+              "Failed to sucessfully stop all outstanding jobs on all "
+              "participants: {} , trying again, retries: {}",
+              test_state,
+              retries);
+        }
+        co_await ss::sleep(1s);
+    }
+    co_return test_state;
+}
+
+ss::future<self_test_frontend::global_test_state> self_test_frontend::status() {
+    return invoke_on_all_nodes(
+      [](auto handle) { return handle->get_status(); });
+}
+
+std::vector<model::node_id>
+self_test_frontend::get_active_participants() const {
+    auto participants = _state.active_participant_ids();
+    auto brokers = participants.empty() ? _members.local().node_ids()
+                                        : participants;
+    return brokers;
+}
+
+template<typename Func>
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::invoke_on_all_nodes(Func f) {
+    auto gate_holder = _gate.hold();
+    auto brokers = get_active_participants();
+    auto local_state_vec = co_await ssx::parallel_transform(
+      brokers, [this, f = std::move(f)](const auto& node_id) {
+          ss::shared_ptr<invoke_wrapper> handle;
+          if (_self == node_id) {
+              handle = ss::make_shared<local_invoke>(node_id, _self_be);
+          } else {
+              handle = ss::make_shared<remote_invoke>(
+                node_id, _self, _connections);
+          }
+          return ss::do_with(
+                   std::move(handle),
+                   [f = std::move(f)](auto& handle) { return f(handle); })
+            .then([node_id](local_test_state result) {
+                return std::pair<const model::node_id, local_test_state>(
+                  node_id, std::move(result));
+            });
+      });
+    auto result = global_test_state(std::move(local_state_vec));
+
+    /// Condense response from all nodes to new known state of the system
+    _state.merge(std::move(result));
+
+    if (!_state.finished()) {
+        co_return _state;
+    }
+    _previous_successful_run = _state;
+    _state = global_test_state({});
+    co_return _previous_successful_run;
```

Comment:
Maybe this will be used in the future, but if not, it seems like this is never used and we might as well not bother keeping track of it?

Reply:
Currently status asks all nodes their status in the case things are finished instead of returning cached result. Wasn't intended, going to return this instead.